File: ruby/lib/task_master/task_repository.rb
module TaskMaster
  class TaskRepository
    def initialize(adapter)
      @adapter = adapter
    end

    def list(sort_by: "Task ID", order: :desc)
      @list ||= @adapter.list.map { |task_data| Task.new(**task_data) }
    end

    def list_by_status(statuses, sort_by: "Task ID", order: :desc)
      list.select { |task| task.status_of?(statuses) }
    end

    def list_by_assignee(assignees, sort_by: "Task ID", order: :desc)
      list.select { |task| task.assigned_to?(assignees) }
    end

    def list_upcoming(sort_by: "Task ID", order: :desc)
      list.select { |task| task.upcoming? }
    end

    def list_overdue(sort_by: "Task ID", order: :desc)
      list.select { |task| task.overdue? }
    end

    def list_completed(sort_by: "Task ID", order: :desc)
      list.select { |task| task.complete? }
    end
  end
end

File: javascript/lib/TaskRepository.js
const Task = require("./Task");

class TaskRepository {
	constructor(adapter) {
		this.adapter = adapter;
	}

	list() {
		return this.adapter.list().map(taskData => {
			return new Task(taskData);
	  })
	}

	listByStatus({statuses, sortBy, order}) {
		console.log(statuses)
		console.log("here")
		return this.list().filter(task => {
			return task.hasStatusOf(statuses) == true;
		})
	}

	listByAssignee({ assignees, sortBy, order }) {
		return this.list().filter(task => {
			return task.isAssignedTo(assignees) == true;
		})
	}

	listUpcoming() {
		return this.list().filter(task => {
			return task.isUpcoming() == true;
		})
	}

	listOverdue() {
		return this.list().filter(task => {
			return task.isOverdue() == true;
		})
	}

	listComplete() {
		return this.list().filter(task => {
			return task.isComplete() == true;
		})
	}
}

module.exports = TaskRepository;
